core/src/proxy/fetch_stage_manager.rs

```diff
@@ -61,97 +73,104 @@ impl FetchStageManager {
-                            }
-                            if fetch_connected && pending_disconnect && pending_disconnect_ts.elapsed() > DISCONNECT_DELAY {
-                                info!("disconnecting fetch stage");
-                                fetch_connected = false;
-                                pending_disconnect = false;
-                                if let Err(e) = Self::set_tpu_addresses(&cluster_info, tpu_addr, tpu_forward_addr) {
-                                    error!("error setting tpu or tpu_fwd to ({:?}, {:?}), error: {:?}", tpu_addr, tpu_forward_addr, e);
-                                }
-                            }
-                        } else {
-                            {
-                                warn!("relayer heartbeat receiver disconnected, shutting down");
-                                return;
+                    heartbeat_received = false;
+                }
+                recv(heartbeat_rx) -> tpu_info => {
+                    if let Ok((tpu_addr, tpu_forward_addr)) = tpu_info {
+                        heartbeats_received += 1;
+                        heartbeat_received = true;
+                        if fetch_connected && !pending_disconnect {
+                            info!("received heartbeat while fetch stage connected, pending disconnect after delay");
+                            pending_disconnect_ts = Instant::now();
+                            pending_disconnect = true;
+                        }
+                        if fetch_connected && pending_disconnect && pending_disconnect_ts.elapsed() > DISCONNECT_DELAY {
+                            info!("disconnecting fetch stage");
+                            fetch_connected = false;
+                            pending_disconnect = false;
+                            if let Err(e) = Self::set_tpu_addresses(cluster_info, tpu_addr, tpu_forward_addr) {
+                                error!("error setting tpu or tpu_fwd to ({:?}, {:?}), error: {:?}", tpu_addr, tpu_forward_addr, e);
                             }
                         }
+                    } else {
+                        {
+                            warn!("relayer heartbeat receiver disconnected, shutting down");
+                            return Err(ProxyError::HeartbeatChannelError);
+                        }
                     }
-                    recv(metrics_tick) -> _ => {
-                        datapoint_info!(
-                            "relayer-heartbeat",
-                            ("fetch_stage_packets_forwarded", packets_forwarded, i64),
-                            ("heartbeats_received", heartbeats_received, i64),
-                        );
-
+                }
+                recv(relayer_config_tick) -> _ => {
+                    if local_relayer_config != global_relayer_config.lock().unwrap().clone() {
+                        return Err(ProxyError::RelayerConfigChanged);
```

Comment:
doesn't seem like this should be an error? wdyt we just update the relevant vars if this changes

Reply:
we use Err to reconnect

Reply:
we use the Err to signal the loop to reconnect. the return type is `()`. wdyt makes more sense?